File: iOS/KinKitchen/KinKitchen/Services/Supabase/DietaryService.swift
//
//  DietaryService.swift
//  KinKitchen
//
//  Created by Greg Hudler on 8/28/26.
//

import Foundation
import Supabase

enum DietaryService {

    static func fetchAllergens() async throws -> [Allergen] {
        try await SupabaseManager.client
            .from("allergens")
            .select()
            .order("name")
            .execute()
            .value
    }

    static func fetchSelectedAllergens() async throws -> [UUID] {
        let user = try await SupabaseManager.client.auth.session.user

        let selections: [UserAllergenSelection] = try await SupabaseManager.client
            .from("user_allergens")
            .select()
            .eq("user_id", value: user.id)
            .execute()
            .value

        return selections.map(\.allergenId)
    }

    static func addAllergen(_ allergenId: UUID) async throws {
        let user = try await SupabaseManager.client.auth.session.user

        let selection = UserAllergenSelection(
            userId: user.id,
            allergenId: allergenId
        )

        try await SupabaseManager.client
            .from("user_allergens")
            .insert(selection)
            .execute()
    }

    static func removeAllergen(_ allergenId: UUID) async throws {
        let user = try await SupabaseManager.client.auth.session.user

        try await SupabaseManager.client
            .from("user_allergens")
            .delete()
            .eq("user_id", value: user.id)
            .eq("allergen_id", value: allergenId)
            .execute()
    }
    
    static func fetchDietaryRestrictions() async throws -> [DietaryRestriction] {
        try await SupabaseManager.client
            .from("dietary_restrictions")
            .select()
            .order("name")
            .execute()
            .value
    }

    static func fetchSelectedDietaryRestrictions() async throws -> [UUID] {
        let user = try await SupabaseManager.client.auth.session.user

        let selections: [UserDietaryRestrictionSelection] = try await SupabaseManager.client
            .from("user_dietary_restrictions")
            .select()
            .eq("user_id", value: user.id)
            .execute()
            .value

        return selections.map(\.restrictionId)
    }

    static func addDietaryRestriction(_ restrictionId: UUID) async throws {
        let user = try await SupabaseManager.client.auth.session.user

        let selection = UserDietaryRestrictionSelection(
            userId: user.id,
            restrictionId: restrictionId
        )

        try await SupabaseManager.client
            .from("user_dietary_restrictions")
            .insert(selection)
            .execute()
    }

    static func removeDietaryRestriction(_ restrictionId: UUID) async throws {
        let user = try await SupabaseManager.client.auth.session.user

        try await SupabaseManager.client
            .from("user_dietary_restrictions")
            .delete()
            .eq("user_id", value: user.id)
            .eq("restriction_id", value: restrictionId)
            .execute()
    }
    
    static func fetchDietaryPreferences() async throws -> [DietaryPreference] {
        try await SupabaseManager.client
            .from("dietary_preferences")
            .select()
            .order("name")
            .execute()
            .value
    }

    static func fetchSelectedDietaryPreferences() async throws -> [UUID] {
        let user = try await SupabaseManager.client.auth.session.user

        let selections: [UserDietaryPreferenceSelection] = try await SupabaseManager.client
            .from("user_dietary_preferences")
            .select()
            .eq("user_id", value: user.id)
            .execute()
            .value

        return selections.map(\.preferenceId)
    }

    static func addDietaryPreference(_ preferenceId: UUID) async throws {
        let user = try await SupabaseManager.client.auth.session.user

        let selection = UserDietaryPreferenceSelection(
            userId: user.id,
            preferenceId: preferenceId
        )

        try await SupabaseManager.client
            .from("user_dietary_preferences")
            .insert(selection)
            .execute()
    }

    static func removeDietaryPreference(_ preferenceId: UUID) async throws {
        let user = try await SupabaseManager.client.auth.session.user

        try await SupabaseManager.client
            .from("user_dietary_preferences")
            .delete()
            .eq("user_id", value: user.id)
            .eq("preference_id", value: preferenceId)
            .execute()
    }
}

private struct UserAllergenSelection: Codable {
    let userId: UUID
    let allergenId: UUID

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case allergenId = "allergen_id"
    }
}

private struct UserDietaryRestrictionSelection: Codable {
    let userId: UUID
    let restrictionId: UUID

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case restrictionId = "restriction_id"
    }
}

private struct UserDietaryPreferenceSelection: Codable {
    let userId: UUID
    let preferenceId: UUID

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case preferenceId = "preference_id"
    }
}

File: Android/app/src/main/java/com/pushtomaindev/kinkitchen/services/supabase/DietaryService.kt
package com.pushtomaindev.kinkitchen.services.supabase

import com.pushtomaindev.kinkitchen.models.user.Allergen
import com.pushtomaindev.kinkitchen.models.user.DietaryPreference
import com.pushtomaindev.kinkitchen.models.user.DietaryRestriction
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Mirrors iOS `DietaryService`. */
object DietaryService {

    private val client get() = SupabaseManager.client

    // MARK: - Allergens

    suspend fun fetchAllergens(): List<Allergen> =
        client.from("allergens")
            .select { order("name", Order.ASCENDING) }
            .decodeList()

    suspend fun fetchSelectedAllergens(): List<String> {
        val userId = requireUserId()
        return client.from("user_allergens")
            .select { filter { eq("user_id", userId) } }
            .decodeList<UserAllergenSelection>()
            .map { it.allergenId }
    }

    suspend fun addAllergen(allergenId: String) {
        client.from("user_allergens")
            .insert(UserAllergenSelection(requireUserId(), allergenId))
    }

    suspend fun removeAllergen(allergenId: String) {
        val userId = requireUserId()
        client.from("user_allergens").delete {
            filter {
                eq("user_id", userId)
                eq("allergen_id", allergenId)
            }
        }
    }

    // MARK: - Dietary Restrictions

    suspend fun fetchDietaryRestrictions(): List<DietaryRestriction> =
        client.from("dietary_restrictions")
            .select { order("name", Order.ASCENDING) }
            .decodeList()

    suspend fun fetchSelectedDietaryRestrictions(): List<String> {
        val userId = requireUserId()
        return client.from("user_dietary_restrictions")
            .select { filter { eq("user_id", userId) } }
            .decodeList<UserDietaryRestrictionSelection>()
            .map { it.restrictionId }
    }

    suspend fun addDietaryRestriction(restrictionId: String) {
        client.from("user_dietary_restrictions")
            .insert(UserDietaryRestrictionSelection(requireUserId(), restrictionId))
    }

    suspend fun removeDietaryRestriction(restrictionId: String) {
        val userId = requireUserId()
        client.from("user_dietary_restrictions").delete {
            filter {
                eq("user_id", userId)
                eq("restriction_id", restrictionId)
            }
        }
    }

    // MARK: - Dietary Preferences

    suspend fun fetchDietaryPreferences(): List<DietaryPreference> =
        client.from("dietary_preferences")
            .select { order("name", Order.ASCENDING) }
            .decodeList()

    suspend fun fetchSelectedDietaryPreferences(): List<String> {
        val userId = requireUserId()
        return client.from("user_dietary_preferences")
            .select { filter { eq("user_id", userId) } }
            .decodeList<UserDietaryPreferenceSelection>()
            .map { it.preferenceId }
    }

    suspend fun addDietaryPreference(preferenceId: String) {
        client.from("user_dietary_preferences")
            .insert(UserDietaryPreferenceSelection(requireUserId(), preferenceId))
    }

    suspend fun removeDietaryPreference(preferenceId: String) {
        val userId = requireUserId()
        client.from("user_dietary_preferences").delete {
            filter {
                eq("user_id", userId)
                eq("preference_id", preferenceId)
            }
        }
    }
}

// Join-table rows. Kept private to this file, matching iOS.

@Serializable
private data class UserAllergenSelection(
    @SerialName("user_id") val userId: String,
    @SerialName("allergen_id") val allergenId: String,
)

@Serializable
private data class UserDietaryRestrictionSelection(
    @SerialName("user_id") val userId: String,
    @SerialName("restriction_id") val restrictionId: String,
)

@Serializable
private data class UserDietaryPreferenceSelection(
    @SerialName("user_id") val userId: String,
    @SerialName("preference_id") val preferenceId: String,
)
